hail/hail/src/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala

```diff
@@ -465,212 +503,212 @@ case class DenseContexts(nRows: TrivialIR, nCols: TrivialIR, contexts: TrivialIR
-    val (newRowPos, newRowIdx, nestedSparsities) =
-      BlockMatrixSparsity.groupedCSCSparsityIR(staticRowPos, staticRowIdx, rowDeps, colDeps)
-    val newContexts = ToArray(flatMapIR(rangeIR(newNCols)) { j =>
-      bindIRs(ArrayRef(newRowPos, j), ArrayRef(newRowPos, j + 1)) { case Seq(start, end) =>
-        mapIR(rangeIR(start, end)) { pos =>
-          IRBuilder.scoped { ib =>
-            val i = ib.memoize(ArrayRef(newRowIdx, pos))
-            val nested = ib.memoize(ArrayRef(nestedSparsities, pos))
-            val nRows = ib.memoize(ArrayRef(rowBlockSizes, i))
-            val nCols = ib.memoize(ArrayRef(colBlockSizes, j))
-            val nestedRowPos = ib.memoize(GetTupleElement(nested, 0))
-            val nestedRowIdx = ib.memoize(GetTupleElement(nested, 1))
-            val nestedContexts = ib.memoize(mapIR(ToStream(GetTupleElement(nested, 2))) { oldPos =>
-              ArrayRef(contexts, oldPos)
-            })
-            SparseContexts(nRows, nCols, nestedRowPos, nestedRowIdx, nestedContexts).irValue
+    val newSparsity = typ.sparsity.asInstanceOf[MatrixSparsity.Sparse]
+    val blockSparsities = newSparsity.definedCoords.map { coords =>
+      sparsity.filter(rowDeps(coords._1), colDeps(coords._2))
+    }
+    val blockSparsitiesTuples = blockSparsities.map { sparsity =>
+      val csc = sparsity.toCSC
+      Row(sparsity.nRows, sparsity.nCols, csc.rowPos, csc.rowIdx)
+    }
+    val blockNewToOld = blockSparsities.map(sparsity.newToOldPos)
+    val blockSparsityType = TStruct(
+      "nRows" -> TInt32,
+      "nCols" -> TInt32,
+      "rowPos" -> TArray(TInt32),
+      "rowIdx" -> TArray(TInt32),
+    )
+    val blockSparsitiesIR = Literal(TArray(blockSparsityType), blockSparsitiesTuples)
+    val blockNewToOldIR = Literal(TArray(TArray(TInt32)), blockNewToOld)
+    val newContexts = ToArray(zipIR(
+      ArraySeq(ToStream(blockSparsitiesIR), ToStream(blockNewToOldIR)),
+      ArrayZipBehavior.AssertSameLength,
+    ) { case Seq(blockSparsity, newToOld) =>
+      InsertFields(
+        blockSparsity,
+        ArraySeq("contexts" -> mapIR(ToStream(newToOld))(ArrayRef(contexts, _))),
+      )
+    })
+
+    new SparseContexts(newSparsity, DynamicSparseContexts(newSparsity, newContexts, ib))
+  }
+
+  override def collect(makeBlock: (Ref, Ref, Ref) => IR): IR =
+    dynamic.collect(makeBlock)
+}
+
+object BlockMatrixStage2 {
+  private def allBlocksColMajor(sparsity: MatrixSparsity): IndexedSeq[(Int, Int)] =
+    sparsity match {
+      case sparse: MatrixSparsity.Sparse =>
+        sparse.definedCoords
+      case MatrixSparsity.Dense(nRows, nCols) =>
+        val foo = Array.fill[(Int, Int)](nRows * nCols)(null)
+        var idx = 0
+        var j = 0
+        while (j < nCols) {
+          var i = 0
+          while (i < nRows) {
+            foo(idx) = i -> j
+            i += 1
+            idx += 1
           }
+          j += 1
         }
-      }
-    })
+        foo
```

Comment:
Does it really have to be so low-level? 
```suggestion
        val coords = new Array[(Int, Int)](nRows * nCols)
        for (j <- 0 until nCols) {
          val k = j * nCols
          for (i <- 0 until nRows) 
            coords(k + i) = i -> j
        }
        ArraySeq.unsafeWrapArray(coords)
```